### PR TITLE
Fix generating export declarations with decorators

### LINKED ISSUE #4585: Exporting decorated class error
### Input Code

``` js
@Decorator()
export class MyClass {}
```
### Babel Configuration (.bablerc, package.json, cli command)

``` js
{
    "plugins": [
        "syntax-decorators"
    ]
}
```
### Expected Behavior

``` js
@Decorator()
export class MyClass {}
```
### Current Behavior

``` js
export @Decorator()
class MyClass {}
```
### Possible Solution

What I'm currently doing to circumvent this:

``` js
export let MyClass = @Decorator()
class MyClass {}
```

¯\\\_(ツ)\_/¯
### Context

I've got a project where I'm trying to parse ES6 code with Flow types, and just take the types out. This minimal case reproduces the same issue, though.
### Your Environment

| software | version |
| --- | --- |
| Babel | babel-cli@6.14.0 |
| node | 6.5.0 |
| npm | 3.10.3 |
| Operating System | Win 10 x64 |

### PR DESCRIPTION
**Pull request by devongovett**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | yes
| Major: Breaking Change?  | no
| Minor: New Feature?      | no
| Deprecations?            | no
| Spec Compliancy?         | no
| Tests Added/Pass?        | yes
| Fixed Tickets            | Fixes #4585
| License                  | MIT
| Doc PR                   | no
| Dependency Changes       | no

Fixes #4585. When running the code generator exporting a class with a decorator, results like the following would be returned prior to this change.

```javascript
export @decorator
class Test {}
```

After this PR:

```javascript
@decorator
export class Test {}
```

### CODE REVIEW

File: packages/babel-generator/src/generators/modules.js [hunk 1]
```diff
@@ -53,13 +53,21 @@ export function ExportAllDeclaration(node: Object) {
...
+  }
+
   this.word("export");
   this.space();
   ExportDeclaration.apply(this, arguments);
 }
 
-export function ExportDefaultDeclaration() {
+export function ExportDefaultDeclaration(node: Object) {
+  if (t.isClassDeclaration(node.declaration)) {
```
Comment (Kovensky) [inline packages/babel-generator/src/generators/modules.js]: This also (almost certainly) affects `FunctionDeclaration` as well, on both the default and named cases.

Comment (devongovett) [inline packages/babel-generator/src/generators/modules.js]: Decorators can only be applied to classes and methods, not free functions.

**Comment by loganfsmyth** [comment]
Am I terrible if I try to take back what I said in https://github.com/babel/babel/issues/4585#issuecomment-250186981?

Having looked at the decorator spec a bit more recently, I'm concerned that the current usage, which this changes us to generate, is actually not accurate according to the spec, with the generator's current output actually being spec-compliant. That would mean the better fix for this would be allowing

```
export @foo class foo {}
```
to parse properly in Babylon.

Thoughts?

**Comment by devongovett** [comment]
Are you thinking of a breaking change such that existing code with the decorator before the export would be invalid, or just changing babylon to accept both?

From the spec [here](https://github.com/wycats/javascript-decorators/blob/master/README.md) it's not clear to me what the expected behavior is when combined with decorators. Did you find something that mentioned this explicitly?

**Comment by devongovett** [comment]
In fact, in at least one example from the spec [here](http://tc39.github.io/proposal-decorators/), the current syntax with the decorator before the export is used.

```javascript
@useShadowDOM
@customElement('my-expander')
export class Expander {
  // ...
}
```

**Comment by devongovett** [comment]
@loganfsmyth thoughts?

**Comment by DrewML** [comment]
@devongovett 

> In fact, in at least one example from the spec here, the current syntax with the decorator before the export is used.

That's just labeled informative though, and isn't part of the spec text itself. It's not unusual for a spec to list prior art, but then deviate from the exact semantics.

Also, I think the spec text you linked to in [this](https://github.com/babel/babel/pull/5020#issuecomment-268723873) comment is outdated (last update Sep 2015, was from strawman proposal). The latest revisions are the one's provided in your second link.

Pinging @wycats to see if he can verify this for sure, but I'm not seeing anything that leads me to believe that `@foo export class abc {}` is supported in the proposed grammar.

I'll try to find it, but I believe there was a comment from Sebastian awhile back during the initial implementation where he mentioned he didn't see it in the spec text, but added support for it believing it was going to be added.

**Comment by devongovett** [comment]
How can we move forward here? Either we can merge this and keep parsing the same, or change parsing to support e.g. `export @foo class foo {}`.

Since babel has supported decorators before the export for a while, there is a lot of code out there written like that, so it would be bad to break parsing for that. So if we change parsing behavior, we'd need to support both.

**Comment by DrewML** [comment]
> so it would be bad to break parsing for that. So if we change parsing behavior, we'd need to support both.

This is the accepted risk that developers make when relying on things that are stage 0. We'd likely do it in a major version, but stage 0 is _highly_ experimental and it should always be assumed that the spec can change drastically as it moves through the stages.

**Comment by mattfysh** [comment]
thanks for the fix @devongovett ... is it possible to merge this, and look into the parsing issue in another PR? cheers

**Comment by hzoo** [comment]
This isn't allowed in Stage 2 decorators anymore so closing

https://github.com/babel/babylon/pull/587/files#diff-cd89c8d2f00940bd0d50120e23c326edR1

> 3.4. export keyword in between of class body and decorators

**Pull request closed without merging**